File: daemon/quotes.go
package main

import (
  "time"
  "fmt"
  "strings"  
  "net/http"       
  "encoding/json" 
  "database/sql"
  "io/ioutil"   
  _ "github.com/go-sql-driver/mysql"
)

type Quotes struct {
  DB *sql.DB
  List []string
}

type QuoteSend struct {
  Type string `json:"type"`
  Timestamp string `json:"timestamp"`
  Data QuoteStruct `json:"data"`
}

type QuotesStruct struct {
  Quotes QuotesQuote `json:"quotes"` 
}

type QuotesQuote struct {
  Quote []QuoteStruct `json"quote"`
}

type QuoteStruct struct {
  Symbol string `json:"symbol"`
  Last json.Number `json:"last"`
  Bid json.Number `json:"bid"`
  Ask json.Number `json:"ask"`
  Description string `json:"description"`
  ChangePercentage json.Number `json:"change_percentage"`    
}

type RateLimitsStruct struct {
  Used string `json:"used"`
  Expire string `json:"expire"`
  Allowed string `json:"allowed"`
  Available string `json:"available"`    
}

//
// Get quotes from Tradier
//
func (t *Quotes) DoQuotes(UsersId int, channel_websocket chan string) {
  
  // Get the tradier API key for this user.
  api_key, _ := t.get_tradier_api_key(UsersId) 
  
  // Open db connection.
  t.start_mysql_connection()
  defer t.DB.Close()   

  // Loop through getting quotes over and over again.
  for { 
    
    // No need to get quotes if the market is closed.
    open := is_market_open();
    if ! open {
      time.Sleep(2 * time.Second)
      continue
    }
      
    // Build a list of symbols we want quotes on
    symbs := t.get_symbols(UsersId)
    
    // Create URL request.
    url := fmt.Sprintf("https://api.tradier.com/v1/markets/quotes?symbols=%s", symbs)
    
    // Setup http client
    client := &http.Client{}    
    
    // Setup api request
    req, _ := http.NewRequest("GET", url, nil)
    req.Header.Set("Accept", "application/json")
    req.Header.Set("Authorization", fmt.Sprint("Bearer ", api_key)) 

    res, err := client.Do(req)
        
    if err != nil {
      println("do_quote: client.Do(req)")
      println(err)
      time.Sleep(time.Second * 3)
      continue     
    }        
    
    // Make sure the api responded with a 200
    if res.StatusCode != 200 {
      time.Sleep(time.Second * 3)
      continue
    }    
       
    // Read the data we got.
    body, err := ioutil.ReadAll(res.Body)
    
    if err != nil {
      println("do_quote: ReadAll")
      println(err)
    }     
  
    // UnJsonize the data. We do this mainly to not send too much information down the wire.
    // Also we send each quote one at a time instead of all the data together 
    // (not really sure why we do this....should be rethought)
    var data QuotesStruct
    json.Unmarshal(body, &data)

    // Loop through and send the quptes.
    for _, quote := range data.Quotes.Quote { 
            
      s := QuoteSend {
        Type: "Quotes:get_quote",
        Timestamp: time.Now().Format("01/02/06 3:04:05 pm"),
        Data: quote}
      
      b, err := json.Marshal(s)

      if err != nil {
        println("do_quote: json.Marshal")
        println(err)
      }       
      
      // Send to the channel for processing.  
      channel_websocket <- string(b)
      
    }
    
    // Close body.
    res.Body.Close()
    
    // Sleep then do it again.
    time.Sleep(time.Second * 2)
  }  
}

//
// Start mysql connection
//
func (t *Quotes) start_mysql_connection() () {
  
  var err error

  cont := fmt.Sprint(Config.Mysql.Username, ":", Config.Mysql.Password, "@/", Config.Mysql.Database, "?charset=utf8")
  t.DB, err = sql.Open("mysql", cont)
  
  if err != nil {
    panic(err)
  }  
  
}

//
// Return the symbols from current orders we have in.
//
func (t *Quotes) get_order_symbols(id int, syb_list *[]string) {
  
  // Update the user object so we can use the latest from the watch list.
  query := fmt.Sprintf("SELECT OrdersSymbol, OrdersLeg1OptionSymbol, OrdersLeg2OptionSymbol, OrdersLeg3OptionSymbol, OrdersLeg4OptionSymbol FROM Orders WHERE (OrdersUpdatedAt > TIMESTAMP(DATE_SUB(NOW(), INTERVAL 2 day)) OR OrdersStatus = 'Open') AND OrdersAccountId=%d", id)
  
  rows, err := t.DB.Query(query)
  
  if err != nil {
    panic(err)
  }
  
  // Loop through the rows of the result
  var OrdersSymbol string
  var OrdersLeg1OptionSymbol string
  var OrdersLeg2OptionSymbol string
  var OrdersLeg3OptionSymbol string
  var OrdersLeg4OptionSymbol string
  
  for rows.Next() {
    err = rows.Scan(&OrdersSymbol, &OrdersLeg1OptionSymbol, &OrdersLeg2OptionSymbol, &OrdersLeg3OptionSymbol, &OrdersLeg4OptionSymbol)
  
    if err != nil {
      panic(err)
    }

    if len(OrdersLeg1OptionSymbol) > 0 {
      *syb_list = append(*syb_list, OrdersLeg1OptionSymbol) 
    }
    
    if len(OrdersLeg2OptionSymbol) > 0 {
      *syb_list = append(*syb_list, OrdersLeg2OptionSymbol) 
    }
    
    if len(OrdersLeg3OptionSymbol) > 0 {
      *syb_list = append(*syb_list, OrdersLeg3OptionSymbol) 
    }
    
    if len(OrdersLeg4OptionSymbol) > 0 {
      *syb_list = append(*syb_list, OrdersLeg4OptionSymbol) 
    }
    
    if len(OrdersSymbol) > 0 {
      *syb_list = append(*syb_list, OrdersSymbol) 
    }                
  } 
    
}

//
// Return the watch list for a particular user
//
func (t *Quotes) get_watch_list(id int, syb_list *[]string) {
  
  // Update the user object so we can use the latest from the watch list.
  query := fmt.Sprintf("SELECT UsersWatchList FROM Users WHERE UsersId=%d", id)
  
  rows, err := t.DB.Query(query)
  
  if err != nil {
    panic(err)
  }
  
  // Loop through the rows of the result
  var UsersWatchList string
  
  for rows.Next() {
    err = rows.Scan(&UsersWatchList)
  
    if err != nil {
      panic(err)
    }
  } 
  
  if err := json.Unmarshal([]byte(UsersWatchList), syb_list); err != nil {
    println("json:", err)
  }   
    
}

//
// Return a list of open positions.
//
func (t *Quotes) get_open_positions(id int, syb_list *[]string) {

  // query
  query := fmt.Sprintf("SELECT SymbolsShort FROM Positions LEFT JOIN Symbols ON PositionsSymbolId = SymbolsId WHERE  PositionsStatus='Open' AND PositionsAccountId=%d", id)
  rows, err := t.DB.Query(query)
  
  if err != nil {
    panic(err)
  }
  
  // Loop through the rows of the result
  var SymbolsShort string  
  
  // Loop through the rows of the result
  for rows.Next() {

    err = rows.Scan(&SymbolsShort)
  
    if err != nil {
      panic(err)
    }
    
    *syb_list = append(*syb_list, SymbolsShort)   
  }

}

//
// Get the tradier API key. If we ever make this a multi tenant app we need to make this smarter.
//
func (t *Quotes) get_tradier_api_key(user_id int) (string, string) {
  
  // Loop through the users in the config to find the key
  for _, row := range Config.Users {
    if row.UsersId == user_id {
      return row.UsersTradierToken, ""
    }
  }
  
  // Return with error
  return "", "key not found"
}

//
// Manage the list of quotes we need to get
//
func (t *Quotes) get_symbols(UsersId int) (string) {
  
  var tmp []string
  var syb_list []string
  var seen = map[string]bool{}

  // Get watchlist symbols
  t.get_watch_list(UsersId, &tmp)
      
  // Get a list of symbols that we currently have positions on.
  t.get_open_positions(UsersId, &tmp)
  
  // Get symbols from current orders.
  t.get_order_symbols(UsersId, &tmp)
  
  // Remove any duplicates
  for _, row := range tmp {
    
    if seen[row] == true {
      continue
    } else {
      seen[row] = true
      syb_list = append(syb_list, row)
    }
    
  } 
  
  // Create a list of CSV symbols to pass to the Tradier API.
  symbs := strings.Join(syb_list, ",") 
 
  // Return CSV of symbols
  return symbs  
  
}

/* End File */

File: app/Http/Controllers/Api/V1/Quotes.php
<?php

namespace App\Http\Controllers\Api\V1;

USE DB;
use App;
use Auth;
use Crypt;
use Input;
use Cache;
use Request;
use DateTime;
use DateInterval;

class Quotes extends \Cloudmanic\LaravelApi\Controller 
{ 
	public $validation_create = [];
	public $validation_update = [];	
	
	//
	// Return quotes for the entire account.
	//
	public function get_account_quotes()
	{
  	$sybs = [];
  	
  	// Get symbols we need from orders
  	$orders = DB::select("SELECT OrdersSymbol, OrdersLeg1OptionSymbol, OrdersLeg2OptionSymbol, OrdersLeg3OptionSymbol, OrdersLeg4OptionSymbol FROM Orders WHERE (OrdersUpdatedAt > TIMESTAMP(DATE_SUB(NOW(), INTERVAL 2 day)) OR OrdersStatus = 'Open') AND OrdersAccountId=" . Auth::user()->UsersId);
  	
    foreach($orders AS $key => $row)
    {
      foreach($row AS $key2 => $row2)
      {
        if(! empty($row2))
        {
          $sybs[] = $row2;
        }
      }
    }
    
    // Get Positions
    $positions = DB::select("SELECT SymbolsShort FROM Positions LEFT JOIN Symbols ON PositionsSymbolId = SymbolsId WHERE  PositionsStatus='Open' AND PositionsAccountId=" . Auth::user()->UsersId);
  	
    foreach($positions AS $key => $row)
    {
      $sybs[] = $row->SymbolsShort;
    }  	

    // Get Watchlist.
    $watch = DB::select("SELECT UsersWatchList FROM Users WHERE UsersId=" . Auth::user()->UsersId);

    foreach(json_decode($watch[0]->UsersWatchList, true) AS $key => $row)
    {
      $sybs[] = strtoupper($row);
    } 
  	
    // Get quotes from Tradier
    $tradier = App::make('App\Library\Tradier');
    $tradier->set_token(Crypt::decrypt(Auth::user()->UsersTradierToken));
    $data = $tradier->get_quotes($sybs);  	
    
  	// Return data.
  	return $this->api_response($data);
	}
	
	//
	// Return current S&P 500 rank
	//
	public function get_snp_500_rank($days)
	{
    $vix_rank = Cache::get('Quotes.SnP500.Rank.' . $days, function() use ($days) {   	
      // Vix.
      $today = 0;
      $vix_rank = 0;
      $vix_running = [];   	
      
      // Get VIX data from Tradier.
      $tradier = App::make('App\Library\Tradier');
      $tradier->set_token(Crypt::decrypt(Auth::user()->UsersTradierToken));
      $data = $tradier->get_historical_pricing('vix', 'daily', date('n/j/Y', strtotime("-$days days")), date('n/j/Y')); 
      
      // Get the current price for the VIX.
      $vix = $tradier->get_quotes([ 'vix' ]); 
      $vix = $vix[0];
      $today = $vix['last'];    
      
      // Loop through data.              
      foreach($data AS $key => $row)
      {
        if(count($vix_running) >= $days)
        {
          array_shift($vix_running);
        } 
      
        $vix_running[] = $row['close'];
      }
      
      // Calculate the IVR
      $l = 0;
      $s = 0;
      
      foreach($vix_running AS $key2 => $row2)
      {
        if($row2 < $today)
        {
          $l++;
        } 
        
        if($row2 == $today)
        {
          $s++;
        } 
        
        $vix_rank = number_format((($l + (0.5 * $s)) / count($vix_running)) * 100, 2);
      }
      
      // Put in cache (5 min cache)
      Cache::put('Quotes.SnP500.Rank.' . $days, $vix_rank, 5);        
      
      // Return rank.
      return $vix_rank;   
    });      
    
    // Return data.
    return $this->api_response([ 'Rank' => $vix_rank ]);
	}
	
	
  //
  // Get timesales data.
  //
  public function timesales()
  {
    // Setup Tradier instance
    $tradier = App::make('App\Library\Tradier');
    $tradier->set_token(Crypt::decrypt(Auth::user()->UsersTradierToken));
    
    // Setup Start times.
    if(is_integer(Input::get('start')))
    {
      $start = date('Y-m-d 09:30', Input::get('start'));
    } else
    {
      $start = date('Y-m-d 09:30', strtotime(Input::get('start')));      
    }

    // Setup End times.
    if(is_integer(Input::get('end')))
    {
      $end = date('Y-m-d 16:00', Input::get('end'));  
    } else
    {
      $end = date('Y-m-d 16:00', strtotime(Input::get('end')));      
    }
    
    // Sometimes we just want to use presets.
    if(Input::get('preset'))
    {
      // Set today.
      $today = new DateTime();      
      
      switch(Input::get('preset'))
      {
        case 'today-1':
          $inv = '1min';
          $start = date('n/j/Y');
          $end = date('n/j/Y');          
        break;
      
        case 'today-5':
          $inv = '5min';
          $start = date('n/j/Y');
          $end = date('n/j/Y');          
        break;
        
        case 'today-15':
          $inv = '15min';
          $start = date('n/j/Y');
          $end = date('n/j/Y');        
        break;
        
        case '5d-15':
          $inv = '15min';
          $start = $today->sub(new DateInterval('P7D'))->format('n/j/Y');
          $end = date('n/j/Y');                
        break;      
        
        case '1y-daily':
          $inv = 'daily';
          $start = $today->sub(new DateInterval('P365D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
        
        case '1y-weekly':
          $inv = 'weekly';
          $start = $today->sub(new DateInterval('P365D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
      
        case '1y-monthly':
          $inv = 'monthly';
          $start = $today->sub(new DateInterval('P365D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
        
        case '2y-daily':
          $inv = 'daily';
          $start = $today->sub(new DateInterval('P730D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
        
        case '2y-weekly':
          $inv = 'weekly';
          $start = $today->sub(new DateInterval('P730D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
      
        case '2y-monthly':
          $inv = 'monthly';
          $start = $today->sub(new DateInterval('P730D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break; 
        
        case '5y-weekly':
          $inv = 'weekly';
          $start = $today->sub(new DateInterval('P1825D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
      
        case '5y-monthly':
          $inv = 'monthly';
          $start = $today->sub(new DateInterval('P1825D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;            
      
        case '10y-weekly':
          $inv = 'weekly';
          $start = $today->sub(new DateInterval('P3650D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break;
      
        case '10y-monthly':
          $inv = 'monthly';
          $start = $today->sub(new DateInterval('P3650D'))->format('n/j/Y');
          $end = date('n/j/Y');        
        break; 
      
        case 'max-monthly':
          $inv = 'monthly';
          $start = '1/1/1980';
          $end = date('n/j/Y');        
        break; 
        
        default:
          return false;
        break;
      }
      
      // Replace input values.
      Input::merge([ 'interval' => $inv ]);
      Input::merge([ 'start' => $start ]);
      Input::merge([ 'end' => $end ]);         
    }    
    
    // Setup Start / End times.
    $start_t = date('Y-m-d 09:30', strtotime($start));
    $end_t = date('Y-m-d 16:00', strtotime($end));     
      
    // Make API call and get data.
    switch(Input::get('interval'))
    {
      case '1min':
      case '5min':
      case '15min':           
        $data = $tradier->get_timesales(Input::get('symbol'), Input::get('interval'), $start_t, $end_t, 'open'); 
      break;
      
      case 'daily':
      case 'weekly':
      case 'monthly':            
        $data = $tradier->get_historical_pricing(Input::get('symbol'), Input::get('interval'), Input::get('start'), Input::get('end'));
        
        // Add timestamps
        foreach($data AS $key => $row)
        {
          $data[$key]['timestamp'] = strtotime($row['date']);
        } 
      break;
      
      default:
        return $this->api_response([]);
      break;      
    }
    
    // Return happy.
    return $this->api_response($data);
  }
}

/* End File */